### ISSUE TITLE
ISOTP/CAN tests are still unstable

### ISSUE THREAD

**Issue by polybassa** [secdev/scapy#3293]
isotp tests are still unstable: https://github.com/secdev/scapy/runs/3061942222?check_suite_focus=true as usual, mostly on PyPy

```
Regression tests for ISOTP
-- Run at 23:43:54 from [test/contrib/isotp.uts] by UTscapy in 57.1609158516
 -> Passed=120
 -> Failed=6

######
## Testing ISOTPSocket with an actual CAN socket
######


###(091)=[failed] Receive frames from CandumpReader

>>> candump_fd = BytesIO(b'''  vcan0  541   [8]  10 0A DE AD BE EF AA AA
...   vcan0  241   [3]  30 00 00
...   vcan0  541   [5]  21 AA AA AA AA
...   vcan0  541   [8]  10 0A DE AD BE EF AA AA
...   vcan0  541   [8]  10 0A DE AD BE EF AA AA
...   vcan0  241   [3]  30 00 00
...   vcan0  541   [5]  21 AA AA AA AA
...   vcan0  541   [8]  10 0A DE AD BE EF AA AA
...   vcan0  241   [3]  30 00 00
...   vcan0  541   [5]  21 AA AA AA AA
...   vcan0  541   [8]  10 0A DE AD BE EF AA AA
...   vcan0  241   [3]  30 00 00
...   vcan0  541   [5]  21 AA AA AA AA
...   vcan0  541   [8]  10 0A DE AD BE EF AA AA
...   vcan0  241   [3]  30 00 00
...   vcan0  541   [5]  21 AA AA AA AA
...   vcan0  541   [8]  10 0A DE AD BE EF AA AA
...   vcan0  241   [3]  30 00 00
...   vcan0  541   [5]  21 AA AA AA AA''')
>>> 
>>> with ISOTPSocket(CandumpReader(candump_fd), sid=0x241, did=0x541, listen_only=True) as s:
...     pkts = s.sniff(timeout=2, count=6)
...     assert(len(pkts) == 6)
...     isotp = pkts[0]
...     print(repr(isotp))
...     print(hex(isotp.src))
...     print(hex(isotp.dst))
...     assert(isotp.data == dhex("DE AD BE EF AA AA AA AA AA AA"))
...     assert(isotp.src == 0x241)
...     assert(isotp.dst == 0x541)
... 
You should put a filter for identifier=541 on your CAN socket
RX state was reset because first frame was received
<ISOTP  data='\xde\xad\xbe\xef\xaa\xaa\xaa\xaa\xaa\xaa' |>
0x241
0x541


###(092)=[failed] Receive frames from CandumpReader with ISOTPSniffer without extended addressing

>>> candump_fd = BytesIO(b'''  vcan0  541   [8]  10 0A DE AD BE EF AA AA
...   vcan0  241   [3]  30 00 00
...   vcan0  541   [5]  21 AA AA AA AA
...   vcan0  541   [8]  10 0A DE AD BE EF AA AA
...   vcan0  541   [8]  10 0A DE AD BE EF AA AA
...   vcan0  241   [3]  30 00 00
...   vcan0  541   [5]  21 AA AA AA AA
...   vcan0  541   [8]  10 0A DE AD BE EF AA AA
...   vcan0  241   [3]  30 00 00
...   vcan0  541   [5]  21 AA AA AA AA
...   vcan0  541   [8]  10 0A DE AD BE EF AA AA
...   vcan0  241   [3]  30 00 00
...   vcan0  541   [5]  21 AA AA AA AA
...   vcan0  541   [8]  10 0A DE AD BE EF AA AA
...   vcan0  241   [3]  30 00 00
...   vcan0  541   [5]  21 AA AA AA AA
...   vcan0  541   [8]  10 0A DE AD BE EF AA AA
...   vcan0  241   [3]  30 00 00
...   vcan0  541   [5]  21 AA AA AA AA''')
>>> 
>>> pkts = sniff(opened_socket=CandumpReader(candump_fd), session=ISOTPSession, timeout=1, session_kwargs={"use_ext_addr": False})
>>> assert(len(pkts) == 6)
>>> isotp = pkts[0]
>>> assert(isotp.data == dhex("DE AD BE EF AA AA AA AA AA AA"))
>>> assert (isotp.dst == 0x541)


###(093)=[failed] Receive frames from CandumpReader with ISOTPSniffer
 all flow control frames are detected as single frame with extended address

>>> candump_fd = BytesIO(b'''  vcan0  541   [8]  10 0A DE AD BE EF AA AA
...   vcan0  241   [3]  30 00 00
...   vcan0  541   [5]  21 AA AA AA AA
...   vcan0  541   [8]  10 0A DE AD BE EF AA AA
...   vcan0  541   [8]  10 0A DE AD BE EF AA AA
...   vcan0  241   [3]  30 00 00
...   vcan0  541   [5]  21 AA AA AA AA
...   vcan0  541   [8]  10 0A DE AD BE EF AA AA
...   vcan0  241   [3]  30 00 00
...   vcan0  541   [5]  21 AA AA AA AA
...   vcan0  541   [8]  10 0A DE AD BE EF AA AA
...   vcan0  241   [3]  30 00 00
...   vcan0  541   [5]  21 AA AA AA AA
...   vcan0  541   [8]  10 0A DE AD BE EF AA AA
...   vcan0  241   [3]  30 00 00
...   vcan0  541   [5]  21 AA AA AA AA
...   vcan0  541   [8]  10 0A DE AD BE EF AA AA
...   vcan0  241   [3]  30 00 00
...   vcan0  541   [5]  21 AA AA AA AA''')
>>> 
>>> pkts = sniff(opened_socket=CandumpReader(candump_fd), session=ISOTPSession, timeout=1)
>>> assert(len(pkts) == 12)
>>> isotp = pkts[1]
>>> assert(isotp.data == dhex("DE AD BE EF AA AA AA AA AA AA"))
>>> assert (isotp.dst == 0x541)
>>> isotp = pkts[0]
>>> assert(isotp.data == dhex(""))
>>> assert (isotp.dst == 0x241)


###(094)=[failed] Receive frames from CandumpReader with ISOTPSniffer and count
 all flow control frames are detected as single frame with extended address

>>> candump_fd = BytesIO(b'''  vcan0  541   [8]  10 0A DE AD BE EF AA AA
...   vcan0  241   [3]  30 00 00
...   vcan0  541   [5]  21 AA AA AA AA
...   vcan0  541   [8]  10 0A DE AD BE EF AA AA
...   vcan0  541   [8]  10 0A DE AD BE EF AA AA
...   vcan0  241   [3]  30 00 00
...   vcan0  541   [5]  21 AA AA AA AA
...   vcan0  541   [8]  10 0A DE AD BE EF AA AA
...   vcan0  241   [3]  30 00 00
...   vcan0  541   [5]  21 AA AA AA AA
...   vcan0  541   [8]  10 0A DE AD BE EF AA AA
...   vcan0  241   [3]  30 00 00
...   vcan0  541   [5]  21 AA AA AA AA
...   vcan0  541   [8]  10 0A DE AD BE EF AA AA
...   vcan0  241   [3]  30 00 00
...   vcan0  541   [5]  21 AA AA AA AA
...   vcan0  541   [8]  10 0A DE AD BE EF AA AA
...   vcan0  241   [3]  30 00 00
...   vcan0  541   [5]  21 AA AA AA AA''')
>>> 
>>> pkts = sniff(opened_socket=CandumpReader(candump_fd), session=ISOTPSession, timeout=1, count=2)
>>> assert(len(pkts) == 2)
>>> isotp = pkts[1]
>>> assert(isotp.data == dhex("DE AD BE EF AA AA AA AA AA AA"))
>>> assert (isotp.dst == 0x541)
>>> isotp = pkts[0]
>>> assert(isotp.data == dhex(""))
>>> assert (isotp.dst == 0x241)


###(095)=[failed] ISOTPSession tests

>>> ses = ISOTPSession()
>>> ses.on_packet_received(None)
>>> ses.on_packet_received([None, None])
>>> assert True


######
## Cleanup
######


###(149)=[failed] Cleanup reference to ISOTPSoftSocket to let the thread end

>>> s = None



UTscapy ended with error code 1

WARNING: UNFINISHED THREADS
[<_MainThread(MainThread, started 140267622429696)>, <Thread(TimeoutScheduler._task, started 140267315259136)>]
ERROR: InvocationError for command /usr/bin/sudo -E .tox/pypy2-linux_root/bin/python -m coverage run -m scapy.tools.UTscapy -c ./test/configs/linux.utsc -K tshark -K icmp_firewall -K not_pypy (exited with code 1)
___________________________________ summary ____________________________________
ERROR:   pypy2-linux_root: commands failed
Error: Process completed with exit code 1.
```

_Originally posted by @gpotter2 in https://github.com/secdev/scapy/issues/3268#issuecomment-879480829_

**Comment by polybassa** [comment]
@gpotter2 Please assignee this one to me.

**Comment by gpotter2** [comment]
Happened again on Windows, exact same error: https://ci.appveyor.com/project/secdev/scapy/builds/39981314/job/mmbd006grktqwewf

**Comment by polybassa** [comment]
Thanks. This gives a little more error information. Will start working on that in the next week probably.

**Comment by gpotter2** [comment]
Thanks, no worries this can wait.

**Comment by polybassa** [comment]
```
######
## ISOTPSoftSocket MITM attack tests
######


###(112)=[failed] Two ISOTPSockets at the same time, multiple sends/receives

>>> with new_can_socket0() as cs1, ISOTPSocket(cs1, sid=0x641, did=0x241) as s1, \
...         new_can_socket0() as cs2, ISOTPSocket(cs2, sid=0x241, did=0x641) as s2:
...     def sender(p):
...         s2.send(p)
...     for i in range(1, 40, 5):
...         isotp = ISOTP(data=bytearray(range(i, i * 2)))
...         t = Thread(target=sender, args=(isotp,))
...         result = s1.sniff(count=1, timeout=5, started_callback=t.start)
...         t.join(timeout=5)
...         assert len(result)
...         assert (result[0].data == isotp.data)
... 
Traceback (most recent call last):
  File "<input>", line 11, in <module>
AssertionError


Exception in thread Thread-166:
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.9.6/x64/lib/python3.9/threading.py", line 973, in _bootstrap_inner
    self.run()
  File "/opt/hostedtoolcache/Python/3.9.6/x64/lib/python3.9/threading.py", line 910, in run
    self._target(*self._args, **self._kwargs)
  File "<input>", line 5, in sender
  File "/home/runner/work/scapy/scapy/scapy/supersocket.py", line 109, in send
    return self.outs.send(sx)
  File "/home/runner/work/scapy/scapy/scapy/contrib/isotp/isotp_soft_socket.py", line 1032, in send
    send_done = self.tx_done.wait(30)
scapy.error.Scapy_Exception: ISOTP send not completed in 30s
```

**Comment by gpotter2** [comment]
So I've had another failure of `cansocket_python` tests and I've started actually looking into those tests.
I don't understand them, nor the one you highlighted above: what are all those threads for? Sending a packet will just add it to the socket's buffer, at least that's what usually happen on windows.. You don't need to have those calls happen simmultaneously.

Running IO things in threads is usually a great way of breaking things, we usually never do that, but apparently that's everything CAN tests are made of. I must say that's our fault for having allowed such tests to get into Scapy, but they should be reworked entirely. I will help upgrade them, starting with `cansocket_python_can`

**Comment by polybassa** [comment]
First of all, I'm very sorry for your trouble with this contribution and thankful for your feedback. 

I'll try to explain the current state of CAN and ISOTP and hopefully answer some questions.

``what are all those threads for? Sending a packet will just add it to the socket's buffer, at least that's what happens on linux sockets.. You do not need those calls to happen simultaneously.``

To fully explain this, I need to split up my answer depending on the used OS and system configuration. For ISOTP I explained this here: https://scapy.readthedocs.io/en/latest/layers/automotive.html#system-compatibilities

For CANSockets, we can have two setups.

Linux, Python3: Such a system can use NativeCANSockets. This is very stable, since most stuff happens in the kernel. As you said, sending a message is just a copy of a message to a kernel socket. Also this tests shouldn't make any problem.

Other Systems or Python2: Here, we can't use Linux SocketCAN directly from Python (or only with a lot of low level code. The guys from python-can did this). To have CANSockets on these systems, we used the python-can project. Unfortunately a python-can Bus object behaves differently than a Socket. To mimic the same runtime behavior as NativeCANSockets have, we implemented an intermediate layer for Scapy. 

Our goal was to have NativeCANSockets and PythonCANSockets behave identical to the user to obtain a platform independent solution. 

ISOTP is a transport-layer on top of CAN. Four special CAN-Frames are used to fragment a bigger ISOTP message into multiple smaller CAN frames. Also, ISOTP has directed communication between two endpoints and some kind of flow control. 

On Linux an Python3 systems, our life is easy. We can use a kernel module, which provides us sockets for ISOTP. These NativeISOTPSockets do all the work in the kernel space.

On all other systems, we need to implement the ISOTP message handling in Python. Since an ISOTPSocket needs to listen for FirstFrame CAN-Messages, which indicate the start of an ISOTP message, we have no other way, as having a thread running and listening on the CAN-Bus. All this is implemented in ISOTPSoftSockets.

Again here, we tried to have identical interfaces on ISOTPSoftSockets and ISOTPNativeSockets.

This is the reason, why we need all this threads.

``Running IO things in threads is usually a great way of breaking things, we usually never do that, but apparently that's everything CAN tests are made of.``

What do you mean with this IO. Does this mean, using `BytesIO` is a bad thing. 


Thank you very much for your help on making this code better. I'm sure, I can improve the current code by using the `ObjectPipe` in PythonCANSockets middle-layer. 

Also I see a lot of improvements for the tests stability and speed by using this `TestSocket` I introduced in #3309 .
This can remove an entire layer of complexity to a lot of test cases. 

I'm almost done with #3054 and luckily I've built these tests not relying on CAN or ISOTP sockets :-) 

After that, I will start with a refactoring of the automotive test cases by using `ObjectPipes` in `TestSockets`

I wish theses tests would fail reliably on my systems, and not only very rarely on CI.

**Comment by gpotter2** [comment]
Thanks a lot for the detailed explanation.

I understand that the soft sockets require a background thread, that is fine. But from what I understood this should allow it to behave like a real socket, meaning there shouldn't need to start some additional threads during the tests.. right?
By IO I meant all the send/receive stuff. The lesser threads we have the safer everything is, unless of course you need one like in this case.

When using `ObjectPipes`, remember you can use `select_objects` on them. This helps a lot keeping everything stable and minimizing the amount of threads required.

Thanks a lot for all of your contributions to Scapy, I really don't want to scare you away from contributing :sweat_smile:

**Comment by polybassa** [comment]
>  understand that the soft sockets require a background thread, that is fine. But from what I understood this should allow it to behave like a real socket, meaning there shouldn't need to start some additional threads during the tests.. right?

Yes, thats right. I will carefully review all disabled tests.

> By IO I meant all the send/receive stuff. The lesser threads we have the safer everything is, unless of course you need one like in this case.

Makes absolute sense. As I said, I've big hopes in the ObjectPipe to remove an entire layer of complexity.

> Thanks a lot for all of your contributions to Scapy, I really don't want to scare you away from contributing sweat_smile

No worries :-D I'm very happy for your feedback. I really want to obtain stable code for this contribution. I'm very happy that you push things forward, all over the place in Scapy. I really like the aim for high quality code in the Scapy project.

**Comment by polybassa** [comment]
@gpotter2 If you have time, I would be happy if some PRs can get merged

**Comment by gpotter2** [comment]
Could you list the order in which you'd like for them to be merged? Thanks

**Comment by polybassa** [comment]
In which I opened the PRs would be ok. Until something conflicts :;)
-- 
Viele Grüße

Nils Weiß

**Comment by Asmeble** [comment]
this needs to be fixed immediately.

**Comment by polybassa** [comment]
#3354 should speed up and stabilize GMLAN Utils tests

**Comment by polybassa** [comment]
#3355 This speeds up and cleans the EcuAnsweringMachine unit tests

**Comment by polybassa** [comment]
I close this for now. I think it's solved after #3487 got merged. If tests fail again, we can track them here.

**Comment by gpotter2** [comment]
Just got an "ECU scan" timeout on the "fix appveyor" PR. Where should we log those?

**Comment by polybassa** [comment]
This error should be solved with #3512. I had a small probability of failing tests, which this PR should remove. 

Any "scanner" related CI-failures are not related to CAN and ISOTP anymore, since I completely decoupled those tests from each other.
